intel_adsp: Fix corrupt fatal logs

teburd opened the pull request:
Adds a lock around the logger backend for adsp and changes the buffer size to 80 to avoid 1 character a time writes to the trace buffer

fixes #44145

andyross reviewed (commented):
I don't think this is the right level for this fix. We shouldn't have "normal" printk and logging racing ever, there's a CONFIG_LOG_PRINTK that should provide a compatible replacement if you need to use both APIs at once (it implements printk() via the logging backend), we shouldn't be mixing backends.

And the proof is that, if you look carefully, this lock added here doesn't actually protect anything. There's already a lock around the winstream, this is just a lock provided for the benefit of other subsystems.

teburd commented:
Even with LOG_PRINTK=y there's garbled output in the example prior to this fix. Maybe not the  best way to fix this and I'm open to suggestions, but the fix does in fact solve a problem

andyross commented:
Then maybe there's missing synchronization in the logging subsystem?

Commit:
logging: backend_adsp: lock the output

Adds a lock around the output to prevent corruption, and sets the buffer
size of the output to size of the output buffer. This fixes the
corrupted log output as process() may be called from multiple contexts
from different CPUs. A background log processing thread may race against
the panic ISR log context also calling process() on an SMP system.

Additionally sets the buffer size to 80 such that the cavs trace_out.c
functionality and lock are useful for more than one character at a time
greatly reducing the chances of garbled output in a printk and log race
but does not entirely prevent it. CONFIG_LOG_PRINTK=y should be used
to avoid all races.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Possibly, the lock around winstream is fine (in trace_out.c) but it effectively is locking character at a time given the way printk works.

I don't disagree this isn't a pretty fix, marking DNM for now, but at least something of a solution (and perhaps a cause) is known now.

teburd commented:
I believe I fully grok the issue(s) now and the commit message reflects all the races I found and how to avoid them, along with a lock around process() in particular avoiding the re-entrance and buffer trampling.

The output looks good.

teburd commented:
There's potentially an alternative fix that involves putting a lock around the body of z_impl_log_process though I'm not sure of all the implications.

nordic-krch commented on subsys/logging/log_backend_adsp.c:
```diff
@@ -10,6 +10,8 @@
 #include <logging/log_output.h>
 #include <logging/log_backend_std.h>
 
+struct k_spinlock lock;
```
static

nordic-krch commented on subsys/logging/log_backend_adsp.c:
```diff
@@ -20,12 +22,10 @@ static int char_out(uint8_t *data, size_t length, void *ctx)
- */
-static uint8_t log_buf[60];
+/* No reason to buffer more, trace out is a buffer */
+static uint8_t log_buf[80];
```
previously it was 60. Any reason for this change?

teburd replied:
I've added a comment and define, also using sizeof where appropriate

nordic-krch commented on subsys/logging/log_backend_adsp.c:
```diff
@@ -20,12 +22,10 @@ static int char_out(uint8_t *data, size_t length, void *ctx)
+static uint8_t log_buf[80];
 
-LOG_OUTPUT_DEFINE(log_output_adsp, char_out, log_buf, 1);
+LOG_OUTPUT_DEFINE(log_output_adsp, char_out, log_buf, 80);
```
sizeof(log_buf) would avoid repeating the value.

nordic-krch commented on subsys/logging/log_backend_adsp.c:
```diff
@@ -76,7 +81,11 @@ static void process(const struct log_backend *const backend,
 {
 	log_format_func_t log_output_func = log_format_func_t_get(log_format_current);
 
+	k_spinlock_key_t key = k_spin_lock(&lock);
```
have you considered to put this lock in `char_out` around `intel_adsp_trace_out`. Here it will lock for significant amount of time in the lower priority thread (in case of deferred mode). Putting it in `char_out` may lead to some interleaving of data within the log message so there is a trade off.

teburd replied:
That was one of the problems being solved here, as on the intel_adsp with a fatal exception ISR being run on one cpu, the lower priority logger thread continues to run and trample the data.

I as debating if a lock should be placed in msg_process in log_core.c instead as it seems there's a race when panic mode and the log thread collide on SMP systems.

kv2019i reviewed (commented):
This seems to address the problem indeed! I did note following weird output when I had also https://github.com/zephyrproject-rtos/zephyr/pull/44529 included. It's still readable, but ordering is unexpected.

@nordic-krch  had good comments, but other than that, looks good to me.

Example log with the backtrace PR also included:
``
535816 zephyr_ll_scheduler_in[00000000] <inf> main: SOF initialized
it(): unsupported domain 2
3552786: SHM: FW ABI 0x3015000 DBG ABI 0x5003000 tags SOF:v2.0-rc1-696-gc9814111ab20-dirty zephyr:zephyr-v3.0.0-1972-g7994080741fb src hash 0x6311e5b4 (ldc hash 0x6311e5b4)
4165080 zephyr_ll_scheduler_init(): unsupported domain 2


Backtrace:0xBE0267B9:[00000030] <err> os:  ** FATAL EXCEPTION
0xBE0761C0 0xBE026A7F:0xBE07[00000030] <err> os:  ** CPU 2 EXCCAUSE 6 (divide by zero)
6210 0xBE02819A:0xBE076240 0x[00000030] <err> os:  **  PC 0xbe0267bc VADDR (nil)
BE0392D4:0xBE076280 0xB[00000030] <err> os:  **  PS 0x60320
E0396C3:0xBE0762A0 0xBE025604:0xBE0762D0 [00000030] <err> os:  **    (INTLEVEL:0 EXCM: 0 UM:1 RING:0 WOE:1 OWB:3 CALLINC:0x2)
BE01557E:0xBE076300 0xBE0137BC:0xBE076330 [00000030] <err> os:  **  A0 0xbe026a82  SP 0xbe0761c0  A2 0x60040000  A3 0x9e07
fc00

[00000030] <err> os:  **  A4 0xc  A5 0x1  A6 0x40000  A7 0xbe02a188
[00000030] <err> os:  **  A8 0xbe02678a  A9 0xbe0761a0 A10 0x9e084740 A11 (nil)
[00000030] <err> os:  ** A12 0x40000 A13 0x2 A14 0x9e07ef78 A15 0x9e07f00c
[00000030] <err> os:  ** LBEG 0xbe040108 LEND 0xbe040116 LCOUNT 0xbe026788
[00000030] <err> os:  ** SAR 0xf
[00000030] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 2
[00000030] <err> os: Current thread: 0x9e07e3b0 (unknown)
[00000030] <err> os: Halting system
``

teburd commented:
> This seems to address the problem indeed! I did note following weird output when I had also #44529 included. It's still readable, but ordering is unexpected.
> 
> @nordic-krch had good comments, but other than that, looks good to me.
> 
> Example log with the backtrace PR also included: `` 535816 zephyr_ll_scheduler_in[00000000] main: SOF initialized it(): unsupported domain 2 3552786: SHM: FW ABI 0x3015000 DBG ABI 0x5003000 tags SOF:v2.0-rc1-696-gc9814111ab20-dirty zephyr:zephyr-v3.0.0-1972-g7994080741fb src hash 0x6311e5b4 (ldc hash 0x6311e5b4) 4165080 zephyr_ll_scheduler_init(): unsupported domain 2
> 
> Backtrace:0xBE0267B9:[00000030] os: ** FATAL EXCEPTION 0xBE0761C0 0xBE026A7F:0xBE07[00000030] os: ** CPU 2 EXCCAUSE 6 (divide by zero) 6210 0xBE02819A:0xBE076240 0x[00000030] os: ** PC 0xbe0267bc VADDR (nil) BE0392D4:0xBE076280 0xB[00000030] os: ** PS 0x60320 E0396C3:0xBE0762A0 0xBE025604:0xBE0762D0 [00000030] os: ** (INTLEVEL:0 EXCM: 0 UM:1 RING:0 WOE:1 OWB:3 CALLINC:0x2) BE01557E:0xBE076300 0xBE0137BC:0xBE076330 [00000030] os: ** A0 0xbe026a82 SP 0xbe0761c0 A2 0x60040000 A3 0x9e07 fc00
> 
> [00000030] os: ** A4 0xc A5 0x1 A6 0x40000 A7 0xbe02a188 [00000030] os: ** A8 0xbe02678a A9 0xbe0761a0 A10 0x9e084740 A11 (nil) [00000030] os: ** A12 0x40000 A13 0x2 A14 0x9e07ef78 A15 0x9e07f00c [00000030] os: ** LBEG 0xbe040108 LEND 0xbe040116 LCOUNT 0xbe026788 [00000030] os: ** SAR 0xf [00000030] os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 2 [00000030] os: Current thread: 0x9e07e3b0 (unknown) [00000030] os: Halting system ``

I believe this is because printk is used to print the backtrace, which can lead to other issues. CONFIG_LOG_PRINTK=y should solve this.

andyross reviewed (dismissed):
This version looks great to me.

andyross commented:
And right: per the discussion above, if there's desire to use printk() in application code alongside the logging macros, we almost certainly want to set CONFIG_LOG_PRINTK=y so that printk() is implemented in terms of logging.  Having both trace subsystems active at once is allowed, but really only useful for special purposes (e.g. you might be debugging one by using the other), and we don't want them interacting more than is absolutely required.  I think interleaving between the two seems acceptable to me given the obvious workaround.

kv2019i commented:
@andyross wrote:
> And right: per the discussion above, if there's desire to use printk() in application code alongside the logging macros, we
> almost certainly want to set CONFIG_LOG_PRINTK=y so that printk() is implemented in terms of logging. Having both trace 

I can confirm this to be the case. With CONFIG_LOG_PRINTK (and this PR), the OS panic is clean including the backtrace. Submitted https://github.com/zephyrproject-rtos/zephyr/pull/44684 to enable this for SOF builds.

andyross reviewed (approved):
Refresh vote
